internal/grpcapi/commit_test.go

```diff
@@ -0,0 +1,216 @@
+)
+
+type CommitTxSuite struct {
+	suite.Suite
+	ctx        context.Context
+	dbConn     *sqlx.DB
+	dockerPool *dockertest.Pool
+	pgResource *dockertest.Resource
+}
+
+func TestCommitTxSuite(t *testing.T) {
+	ts := new(CommitTxSuite)
+	suite.Run(t, ts)
+}
+
+func (ts *CommitTxSuite) SetupSuite() {
+	ctx := context.Background()
+	logger := logging.GetLogger()
+	ctx = logging.WithContext(ctx, logger)
+
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		log.Fatalf("Could not connect to docker: %s", err)
+	}
+
+	dbUser := "user_db"
+	dbName := "users_db"
+	sslMode := "disable"
+	// pulls an image, creates a container based on it and runs it
+	pgResource, err := pool.RunWithOptions(&dockertest.RunOptions{
+		Repository: "postgres",
+		Tag:        "14.2",
+		Env: []string{
+			fmt.Sprintf("POSTGRES_USER=%s", dbUser),
+			fmt.Sprintf("POSTGRES_DB=%s", dbName),
+			"POSTGRES_HOST_AUTH_METHOD=trust",
+			"listen_addresses = '*'",
+		},
+		Cmd: []string{"postgres", "-c", "log_statement=all", "-c", "log_destination=stderr", "--max_prepared_transactions=100"},
+	}, func(config *docker.HostConfig) {
+		// set AutoRemove to true so that stopped container goes away by itself
+		config.AutoRemove = true
+		config.RestartPolicy = docker.RestartPolicy{Name: "no"}
+	})
+	if err != nil {
+		log.Fatalf("Could not start pgResource: %s", err)
+	}
+
+	hostAndPort := pgResource.GetHostPort("5432/tcp")
+	dbHost, dbPort, err := net.SplitHostPort(hostAndPort)
+	if err != nil {
+		log.Fatalf("split host-port '%s' failed: '%s'", hostAndPort, err)
+	}
+	dbConf := &config.DB{
+		ConnString:       fmt.Sprintf("host=%s port=%s user=%s dbname=%s sslmode=%s", dbHost, dbPort, dbUser, dbName, sslMode),
+		MigrationDirPath: "../../sql-migrations",
+		MigrationTable:   "migrations",
+		MaxOpenConns:     20,
+		ConnMaxLifetime:  10 * time.Second,
+	}
+
+	var dbConn *sqlx.DB
+	pool.MaxWait = 30 * time.Second
+	if err = pool.Retry(func() error {
+		dbConn, err = db.Connect(ctx, dbConf)
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		log.Fatalf("Could not connect to docker: %s", err)
+	}
+
+	err = migrations.Apply(ctx, dbConf)
+	if err != nil {
+		log.Fatalf("apply migrations failed: '%s'", err)
+	}
+
+	ts.ctx = ctx
+	ts.pgResource = pgResource
+	ts.dockerPool = pool
+	ts.dbConn = dbConn
+}
+
+func (ts *CommitTxSuite) TearDownSuite() {
+	ctx := ts.ctx
+	dbConn := ts.dbConn
+	if disconnectErr := db.Disconnect(ctx, dbConn); disconnectErr != nil {
+		log.Fatalf("disconnect failed: '%s'", disconnectErr)
+	}
+
+	dockerPool := ts.dockerPool
+	pgResource := ts.pgResource
+	if purgeErr := dockerPool.Purge(pgResource); purgeErr != nil {
+		log.Fatalf("Could not purge pgResource: %s", purgeErr)
+	}
+}
+
+func (ts *CommitTxSuite) TestCommit_OK() {
+	t := ts.T()
+	ctx := ts.ctx
+	logger := logging.FromContext(ctx)
+	dbConn := ts.dbConn
+
+	newTxIDCh := make(chan string, 1)
+	cancelTxIDCh := make(chan string, 1)
+	grpcServer := grpc.NewServer(
```

Comment:
Is the server not a common resource? It may be better to store the connection to the server in the test suite

Reply:
i declined such solution, because:

- i need to check newTxIDCh or cancelTxIDCh in the tests (read the txIDs and assert them)
- i was too lazy to refactor test suite